### PR TITLE
[SPARK-6352] [SQL] Custom parquet output committer

### PR DESCRIPTION
**Pull request by ypcat**
Add new config "spark.sql.parquet.output.committer.class" to allow custom parquet output committer and an output committer class specific to use on s3.
Fix compilation error introduced by https://github.com/apache/spark/pull/5042.
Respect ParquetOutputFormat.ENABLE_JOB_SUMMARY flag.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

**Comment by ypcat** [comment]
@liancheng 1.4.0 will freeze at May 1st, does this PR has chance to get in?

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/DirectParquetOutputCommitter.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            }
+          }
+        }
+      }
+      if (configuration.getBoolean("mapreduce.fileoutputcommitter.marksuccessfuljobs", true)) {
+        val successPath = new Path(outputPath, FileOutputCommitter.SUCCEEDED_FILE_NAME)
+        fileSystem.create(successPath).close()
+      }
+    } catch {
+      case e: Exception => LOG.warn("could not write summary file for " + outputPath, e)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/DirectParquetOutputCommitter.scala]: Exception caught here may be also caused by writing the `_SUCCEEDED` mark file. Can we move the outer most `try`  into the first `if` block, and add a separate `try` in the second `if` block for writing the `_SUCCEEDED` mark file?

**Comment by liancheng** [comment]
@ypcat Sorry for the late reply. This LGTM except for a minor issue.

**Comment by liancheng** [comment]
ok to test

**Comment by liancheng** [comment]
add to whitelist

**Comment by liancheng** [comment]
Thanks for working on this! Merging to master.

**Comment by ypcat** [comment]
cool

**Pull request closed without merging**